In [43]:
import pandas as pd
from sklearn.model_selection import train_test_split
# df = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/'
#                  'mushroom/agaricus-lepiota.data', header=None, engine='python')
df = pd.read_csv('agaricus-lepiota.data.csv')
column_name = ['classes','cap-shape', 'cap-surface','cap-color','bruises?','odor',
               'gill-attachment','gill-spacing','gill-size','gill-color',
               'stalk-shape','stalk-root','stalk-surface-above-ring',
               'stalk-surface-below-ring','stalk-color-above-ring',
               'stalk-color-below-ring','veil-type','veil-color','ring-number',
               'ring-type','spore-print-color','population','habitat']
df.columns = column_name
df.head()

classes cap-shape cap-surface cap-color bruises? odor gill-attachment  \
0       e         x           s         y        t    a               f   
1       e         b           s         w        t    l               f   
2       p         x           y         w        t    p               f   
3       e         x           s         g        f    n               f   
4       e         x           y         y        t    a               f   

  gill-spacing gill-size gill-color   ...   stalk-surface-below-ring  \
0            c         b          k   ...                          s   
1            c         b          n   ...                          s   
2            c         n          n   ...                          s   
3            w         b          k   ...                          s   
4            c         b          n   ...                          s   

  stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
0                      w                      w         p          w   
1                      w                      w         p          w   
2                      w                      w         p          w   
3                      w                      w         p          w   
4                      w                      w         p          w   

  ring-number ring-type spore-print-color population habitat  
0           o         p                 n          n       g  
1           o         p                 n          n       m  
2           o         p                 k          s       u  
3           o         e                 n          a       g  
4           o         p                 k          n       g  

[5 rows x 23 columns]

In [51]:
from sklearn.preprocessing import LabelEncoder
import numpy as np
# encode label first
label_le = LabelEncoder()
df['classes'] = label_le.fit_transform(df['classes'].values)

# encode categorical features

catego_le = LabelEncoder()

# transform categorical values into numerical values
# be careful that '?' will also be encoded
# we have to replace it to NaN in numerical
num_values = []
for i in column_name[1:]:
    df[i] = catego_le.fit_transform(df[i].values)
    classes_list = catego_le.classes_.tolist()
    
    # store the total number of values
    num_values.append(len(classes_list))
    
    # replace '?' with 'NaN'
    if '?' in classes_list:
        idx = classes_list.index('?')
        df[i] = df[i].replace(idx, np.nan)

display(df.head(15))

classes  cap-shape  cap-surface  cap-color  bruises?  odor  \
0         0          5            2          9         1     0   
1         0          0            2          8         1     3   
2         1          5            3          8         1     6   
3         0          5            2          3         0     5   
4         0          5            3          9         1     0   
5         0          0            2          8         1     0   
6         0          0            3          8         1     3   
7         1          5            3          8         1     6   
8         0          0            2          9         1     0   
9         0          5            3          9         1     3   
10        0          5            3          9         1     0   
11        0          0            2          9         1     0   
12        1          5            3          8         1     6   
13        0          5            0          4         0     5   
14        0          4            0          3         0     5   

    gill-attachment  gill-spacing  gill-size  gill-color   ...     \
0                 1             0          0           4   ...      
1                 1             0          0           5   ...      
2                 1             0          1           5   ...      
3                 1             1          0           4   ...      
4                 1             0          0           5   ...      
5                 1             0          0           2   ...      
6                 1             0          0           5   ...      
7                 1             0          1           7   ...      
8                 1             0          0           2   ...      
9                 1             0          0           2   ...      
10                1             0          0           5   ...      
11                1             0          0          10   ...      
12                1             0          1           4   ...      
13                1             1          0           5   ...      
14                1             0          1           4   ...      

    stalk-surface-below-ring  stalk-color-above-ring  stalk-color-below-ring  \
0                          2                       7                       7   
1                          2                       7                       7   
2                          2                       7                       7   
3                          2                       7                       7   
4                          2                       7                       7   
5                          2                       7                       7   
6                          2                       7                       7   
7                          2                       7                       7   
8                          2                       7                       7   
9                          2                       7                       7   
10                         2                       7                       7   
11                         2                       7                       7   
12                         2                       7                       7   
13                         0                       7                       7   
14                         2                       7                       7   

    veil-type  veil-color  ring-number  ring-type  spore-print-color  \
0           0           2            1          4                  3   
1           0           2            1          4                  3   
2           0           2            1          4                  2   
3           0           2            1          0                  3   
4           0           2            1          4                  2   
5           0           2            1          4                  2   
6           0           2            1          4        

In [52]:
# Todo
# deal missing value denoted by '?'
from sklearn.preprocessing import Imputer
from sklearn.model_selection import train_test_split
imr = Imputer(missing_values='NaN', strategy='most_frequent', axis=0)
imr = imr.fit(df.values)
imputed_data = imr.transform(df.values)

df_impute = pd.DataFrame(imputed_data)
df_impute.columns = df.columns

X = df.drop(['classes'], axis=1).values

Y = df['classes'].values
X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.2)

print('done')

done


In [59]:
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import StandardScaler
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score

# find the index of the categorical feature
catego_features_idx = []
for fea in column_name[1:]:
    catego_features_idx.append(df.columns.tolist().index(fea)-1)

print(catego_features_idx)
print(X_train.shape,y_train.shape)
# define pipeline with an arbitrary number of transformer in a tuple array
pipe_knn = Pipeline([('imr', Imputer(missing_values='NaN', strategy='most_frequent', axis=0)),
                     ('ohe', OneHotEncoder(categorical_features=catego_features_idx, 
                                           n_values=num_values, sparse=False)),
                     ('scl', StandardScaler()),
                     ('clf', KNeighborsClassifier(n_neighbors=10, p=2, metric='minkowski'))])

pipe_svm = Pipeline([('imr', Imputer(missing_values='NaN', strategy='most_frequent', axis=0)),
                     ('ohe', OneHotEncoder(categorical_features=catego_features_idx, 
                                           n_values=num_values, sparse=False)),
                     ('scl', StandardScaler()),
                     ('clf', SVC(kernel='rbf', random_state=0, gamma=0.001, C=10.0))])

# use the pipeline model to train
pipe_knn.fit(X_train, y_train)
y_pred = pipe_knn.predict(X_test)
print('[KNN]')
print('Misclassified samples: %d' % (y_test != y_pred).sum())
print('Accuracy: %.4f' % accuracy_score(y_test, y_pred))

pipe_svm.fit(X_train, y_train)
y_pred = pipe_svm.predict(X_test)
print('\n[SVC]')
print('Misclassified samples: %d' % (y_test != y_pred).sum())
print('Accuracy: %.4f' % accuracy_score(y_test, y_pred))

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21]
(6498, 22) (6498,)
[KNN]
Misclassified samples: 354
Accuracy: 0.7822

[SVC]
Misclassified samples: 339
Accuracy: 0.7914


In [57]:
pipe_svm = Pipeline([('imr', Imputer(missing_values='NaN', strategy='most_frequent', axis=0)),
                     ('scl', StandardScaler()),
                     ('clf', SVC(kernel='rbf', random_state=0, gamma=0.001, C=10.0))])

pipe_svm.fit(X_train, y_train)
y_pred = pipe_svm.predict(X_test)
print('\n[SVC]')
print('Misclassified samples: %d' % (y_test != y_pred).sum())
print('Accuracy: %.4f' % accuracy_score(y_test, y_pred))


[SVC]
Misclassified samples: 65
Accuracy: 0.9600


In [58]:
pipe_knn = Pipeline([('imr', Imputer(missing_values='NaN', strategy='most_frequent', axis=0)),
                     ('scl', StandardScaler()),
                     ('clf', KNeighborsClassifier(n_neighbors=10, p=2, metric='minkowski'))])
# use the pipeline model to train
pipe_knn.fit(X_train, y_train)
y_pred = pipe_knn.predict(X_test)
print('[KNN]')
print('Misclassified samples: %d' % (y_test != y_pred).sum())
print('Accuracy: %.4f' % accuracy_score(y_test, y_pred))

[KNN]
Misclassified samples: 2
Accuracy: 0.9988



### 2.What you have done for data preprocessing
->Imputing "?" value in the dataset by setting it to the most frequent values

->Standard scaling
### 3.Your code and accuracy by using KNN and/or SVM
Code: see above 

KNN best acc: 0.9988, SVM: 0.96

### 4.Anything you want to tell us

Onehot encoding doesn't imply better result 

(accuracy with one hot encoding:0.78,without:0.9988 for KNN)

(accuracy with one hot encoding:0.79 (C=10),without:0.96 for SVM)


